game-core/src/system/ally.rs

```diff
@@ -1,19 +1,80 @@
-    core::cgmath::InnerSpace,
+    core::cgmath::{InnerSpace, Vector3},
     core::{Parent, Transform},
     ecs::{Entities, Join, Read, ReadStorage, System, WriteStorage},
     renderer::{SpriteRender, Transparent},
 };
-use crate::component::{Ally, Animation, Player};
+use crate::component::{Ally, Animation, Motion, Player};
 use rand::distributions::{Distribution, Uniform};
 
 pub struct Movement;
 
 impl<'s> System<'s> for Movement {
-    type SystemData = (ReadStorage<'s, Ally>, WriteStorage<'s, Transform>);
+    type SystemData = (
+        ReadStorage<'s, Ally>,
+        WriteStorage<'s, Motion>,
+        ReadStorage<'s, Transform>,
+        ReadStorage<'s, Player>,
+        Entities<'s>,
+    );
+
+    fn run(&mut self, (allies, mut motions, transforms, players, entities): Self::SystemData) {
+        let speed = 20.0;
+        let follow_distance = 30.0;
+        let max_distance = 100.0;
+
+        let p_transform = {
+            let (t, _) = (&transforms, &players)
+                .join()
+                .next()
+                .expect("no player found");
+            (t.clone())
+        };
+
+        for (_, motion, transform1, entity1) in
+            (&allies, &mut motions, &transforms, &entities).join()
+        {
+            let d = p_transform.translation - transform1.translation;
+            let m = d.truncate().magnitude();
+
+            let mut pv = d / m; //unit vector
+                                //less than follow distance, do nothing
+            if m < follow_distance {
+                pv *= 0.0;
+            } else if (m > follow_distance) && (m < max_distance) {
+                //at or more than the follow distance, play a bit of catchup
+                pv *= (1.0 + (m - follow_distance) / (max_distance - follow_distance)) * speed;
+            } else if m > max_distance {
+                //max catchup is 2.0
+                pv *= 2.0 * speed;
+            }
+
+            let mut av = Vector3::new(0.0, 0.0, 0.0);
+
+            for (_, transform2, entity2) in (&allies, &transforms, &entities).join() {
+                if entity1 != entity2 {
+                    let d = transform1.translation - transform2.translation;
+                    let m = d.truncate().magnitude();
+                    let mut v = Vector3::new(1.0, 0.0, 0.0);;
+
+                    if m > 0.0 {
+                        v = d / m;
+                    }
+
+                    let min_distance = 5.0;
 
-    fn run(&mut self, (allies, mut transforms): Self::SystemData) {
-        for (_, _transform) in (&allies, &mut transforms).join() {}
+                    if m < min_distance {
+                        v *= ((min_distance - m) / min_distance) * -1.0 * speed;
+                    } else if (m > follow_distance) && (m < max_distance) {
+                        v *= (1.0 + (m - follow_distance) / (max_distance - follow_distance))
+                            * speed;
+                    }
+
+                    av += v;
```

Comment:
am i reading this correctly in that av accumulates the motion vectors from each entity that is within the necessary distance from it? So if there are lots of allies near you you'll end up with a motion vector of a greater magnitude?

Reply:
The idea is that the allies are moving to a place where the movement is minimized. so, the vectors should start approaching zero. If there are many allies close and they are spread out, the vectors will cancel to a degree.